### PR TITLE
[Merged by Bors] - feat: introduce a new Fin.append and Fin.repeat

### PR DESCRIPTION
**Pull request by Ruben-VandeVelde**
Port of https://github.com/leanprover-community/mathlib/pull/10346

---

Question: should `repeat` be `replicate` instead?

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by eric-wieser** [comment]
I thought about the repeat / replicate thing; but decided against it since `fin.repeat` is akin to `(list.replicate _ _).join` (there is an open mathlib 3 PR with the lemma making this connection)

File: Mathlib/Data/Fin/Tuple/Basic.lean [hunk 1]
```diff
@@ -284,21 +287,119 @@ theorem range_cons {α : Type _} {n : ℕ} (x : α) (b : Fin n → α) :
...
+  · rw [append_right]
+    simp [← natAdd_natAdd]
+#align fin.append_assoc Fin.append_assoc
+
+end Append
+
+section Repeat
+
+/-- Repeat `a` `m` times. For example `Fin.repeat 2 ![0, 3, 7] = ![0, 3, 7, 0, 3, 7]`. -/
+@[simp]
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/Fin/Tuple/Basic.lean:360]:
```suggestion
-- Porting note: removed @[simp]
```

File: Mathlib/Data/Fin/Tuple/Basic.lean [hunk 2]
```diff
@@ -284,21 +287,119 @@ theorem range_cons {α : Type _} {n : ℕ} (x : α) (b : Fin n → α) :
...
+  · rw [append_right]
+    simp [← natAdd_natAdd]
+#align fin.append_assoc Fin.append_assoc
+
+end Append
+
+section Repeat
+
+/-- Repeat `a` `m` times. For example `Fin.repeat 2 ![0, 3, 7] = ![0, 3, 7, 0, 3, 7]`. -/
+-- Porting note: removed @[simp]
```
Comment (eric-wieser) [inline Mathlib/Data/Fin/Tuple/Basic.lean:360]: What went wrong if you kept it?

Comment (Ruben-VandeVelde) [reply Mathlib/Data/Fin/Tuple/Basic.lean:360]: "simp simplifies LHS" at https://github.com/leanprover-community/mathlib4/actions/runs/3947414172/jobs/6756204194

Comment (eric-wieser) [reply Mathlib/Data/Fin/Tuple/Basic.lean:360]: In lean 3, `@[simp]` on an equation compiler definition tagged each of the equations with `simp`.

If that's not the case any more, then we should add a new `repeat_apply` lemma with the same statement.

Comment (eric-wieser) [reply Mathlib/Data/Fin/Tuple/Basic.lean:360]: [Zulip thread](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/.40.5Bsimp.5D.20on.20equation.20compiler.20definitions/near/322022846)

Comment (eric-wieser) [reply Mathlib/Data/Fin/Tuple/Basic.lean:360]: For now, let's add `repeat_apply` and include `leanprover/lean4#2042` in the porting note.

File: Mathlib/Data/Fin/Tuple/Basic.lean [hunk 3]
```diff
@@ -284,21 +287,119 @@ theorem range_cons {α : Type _} {n : ℕ} (x : α) (b : Fin n → α) :
...
+def «repeat» {α : Type _} (m : ℕ) (a : Fin n → α) : Fin (m * n) → α
+  | i => a i.modNat
+#align fin.repeat Fin.repeat
```
Comment (eric-wieser) [inline Mathlib/Data/Fin/Tuple/Basic.lean:363]:
```suggestion
@[simp]
def «repeat» {α : Type _} (m : ℕ) (a : Fin n → α) : Fin (m * n) → α
  | i => a i.modNat
#align fin.repeat Fin.repeat

-- porting note: workaround for leanprover/lean4#2042
theorem «repeat»._sunfold : True := trivial
```

Comment (eric-wieser) [reply Mathlib/Data/Fin/Tuple/Basic.lean:363]: Nevermind, this doesn't work?

**Comment by jcommelin** [comment]
Thanks :tada:

bors merge

**Pull request closed without merging**